# Puzzle

https://thefiddler.substack.com/p/can-you-spot-the-sheep

### This Week’s Fiddler

Two sheep are at two random points inside a square pen. They are munching grass and staring in two random directions. Each sheep has a field of view that’s 180 degrees.

What is the probability that they both see each other?

### This Week’s Extra Credit

Now, three sheep are at three random points inside a square pen. They are munching grass and staring in three random directions. As before, each sheep has a field of view that’s 180 degrees.

What is the probability that all three sheep see each other?

# Fiddler Solution

This could get hairy if one attempted it with certain approaches, but it becomes simple with a different approach.

Let's consider some one fixed position of the 2 sheep, so that the only variables are which directions they are staring in.

Sheep A has probability 1/2 of looking such that it can see sheep B.

And similarly, sheep B has probability 1/2 of looking such that it can see sheep A.

So, the probability they are both looking at each other is 1/4.

Now, we did not consider any information specific to the position, so this holds for all positions.

So, the probability averaged over all positions is also 1/4.

# Extra Credit Solution

## Fixed positions

Let's again start with fixed positions. Let the sheep be A,B,C, forming triangle ABC with angles a,b,c (in radians)

The probability of sheep A looking such that it see both B and C will be $\frac{\pi - a}{2\pi}$, and similarly for sheep B and C.

So, the probability that they all can see each other is:

$P = \frac{(\pi - a)(\pi - b)(\pi - c)}{8\pi^3}$

Since $a + b + c = \pi$, this can be simplified to:

$P =\frac{(\pi - a)(\pi - b)(a+b)}{8\pi^3}$


## Digression - Uniform independent angles.

Now let's consider a slightly different problem where angles a,b are chosen independently and uniformly to construct triangle ABC. Now the problem of finding the probability becomes a lot easier.

$P = \int_{0}^{\pi} da \int_{0}^{\pi-a} db \frac{(\pi - a)(\pi - b)(a+b)}{8\pi^3}  /  \int_{0}^{\pi} da \int_{0}^{\pi-a} db $

I used wolfram alpha to do this integral.

[Full-Integral](https://www.wolframalpha.com/input?i2d=true&i=Divide%5BIntegrate%5BIntegrate%5BDivide%5B%5C%2840%29pi-a%5C%2841%29%5C%2840%29pi-b%5C%2841%29%5C%2840%29a%2Bb%5C%2841%29%2C8Power%5Bpi%2C3%5D%5D%2C%7Bb%2C0%2Cpi+-+a%7D%5D%2C%7Ba%2C0%2Cpi%7D%5D%2CIntegrate%5BIntegrate%5B1%2C%7Bb%2C0%2Cpi-a%7D%5D%2C%7Ba%2C0%2Cpi%7D%5D%5D) = 
[Top-Part](https://www.wolframalpha.com/input?i2d=true&i=Integrate%5BIntegrate%5BDivide%5B%5C%2840%29pi-a%5C%2841%29%5C%2840%29pi-b%5C%2841%29%5C%2840%29a%2Bb%5C%2841%29%2C8Power%5Bpi%2C3%5D%5D%2C%7Bb%2C0%2Cpi+-+a%7D%5D%2C%7Ba%2C0%2Cpi%7D%5D) / 
[Bottom-Part](https://www.wolframalpha.com/input?i2d=true&i=Integrate%5BIntegrate%5B1%2C%7Bb%2C0%2Cpi-a%7D%5D%2C%7Ba%2C0%2Cpi%7D%5D)

And got 

$ P = 7/240 = 0.029167 $

Now, while picking points uniformly within a square is not the same as picking angles uniformly, this gives us an interesting reference.

Also, I am not sure if the case of a square would have higher or lower probability. Maybe higher because acute angles seem more likely, which would raise the probability, but that could also be a limitation of my imagination.


## Numerical approach

In [93]:
from math import acos, sqrt, pi
def three_angles_from_3_points(p1, p2, p3):
    v12 = (p2[0] - p1[0], p2[1] - p1[1])
    v13 = (p3[0] - p1[0], p3[1] - p1[1])
    v23 = (p3[0] - p2[0], p3[1] - p2[1])
    v21 = (-v12[0], -v12[1])
    v31 = (-v13[0], -v13[1])
    v32 = (-v23[0], -v23[1])
    magnitude_v12 = sqrt(v12[0] ** 2 + v12[1] ** 2)
    magnitude_v13 = sqrt(v13[0] ** 2 + v13[1] ** 2)
    magnitude_v23 = sqrt(v23[0] ** 2 + v23[1] ** 2)
    if magnitude_v12 == 0 or magnitude_v13 == 0 or magnitude_v23 == 0:
        return 0, pi/2, pi/2
    dot_product_1 = v12[0] * v13[0] + v12[1] * v13[1]
    dot_product_2 = v21[0] * v23[0] + v21[1] * v23[1]
    #dot_product_3 = v31[0] * v32[0] + v31[1] * v32[1]
    cosine1 = dot_product_1 / (magnitude_v12 * magnitude_v13)
    cosine2 = dot_product_2 / (magnitude_v12 * magnitude_v23)
    #cosine3 = dot_product_3 / (magnitude_v13 * magnitude_v23)
    #print(f"cosine1: {cosine1}, cosine2: {cosine2}")
    if cosine1 < -1: cosine1 = -1
    if cosine1 >  1: cosine1 =  1
    if cosine2 < -1: cosine2 = -1
    if cosine2 >  1: cosine2 =  1
    angle1 = acos(cosine1)
    angle2 = acos(cosine2)
    #angle3 = acos(cosine3)
    angle3 = pi - angle1 - angle2 # faster than acos(cosine3)
    return angle1, angle2, angle3

print (three_angles_from_3_points((0, 0), (1, 0), (1, 1)))
print (three_angles_from_3_points((0, 0), (1, 0), (2, 2)))
print (three_angles_from_3_points((0, 0), (1, 1), (2, 2)))
print (three_angles_from_3_points((1, 1), (1, 1), (2, 2)))
print (three_angles_from_3_points((1, 1), (1, 1), (1, 1)))

import time
import random
total_time = 0
for _ in range(1):
    p1 = (random.uniform(-10, 10), random.uniform(-10, 10))
    p2 = (random.uniform(-10, 10), random.uniform(-10, 10))
    p3 = (random.uniform(-10, 10), random.uniform(-10, 10))
    start_time = time.time()
    three_angles_from_3_points((0, 0), (1, 0), (1, 1))
    total_time += time.time() - start_time
print ("total_time: ", total_time)


(0.7853981633974484, 1.5707963267948966, 0.7853981633974483)
(0.7853981633974484, 2.0344439357957027, 0.32175055439664213)
(2.1073424255447017e-08, 3.1415926325163688, 0.0)
(0, 1.5707963267948966, 1.5707963267948966)
(0, 1.5707963267948966, 1.5707963267948966)
total_time:  0.0


In [94]:
from math import pi
def prob_all_sheep_see_each_other(xa,ya,xb,yb,xc,yc):
    p1 = (xa, ya)
    p2 = (xb, yb)
    p3 = (xc, yc)

    a, b, c = three_angles_from_3_points(p1, p2, p3)

    p1 = (pi - a) / (2 * pi)
    p2 = (pi - b) / (2 * pi)
    p3 = (pi - c) / (2 * pi)

    combined_probability = p1 * p2 * p3
    return combined_probability

def constant_one(xa,ya,xb,yb,xc,yc):
    return 1

In [95]:
from scipy.integrate import nquad

def my_integrate(fn, epsilon):
    lower_bounds = [0, 0, 0, 0, 0, 0]
    upper_bounds = [1, 1, 1, 1, 1, 1]
    ranges = list(zip(lower_bounds, upper_bounds))

    result = nquad(fn, ranges, opts={'epsabs': epsilon, 'epsrel': epsilon},  full_output=True)
    return result

In [97]:
my_integrate(constant_one, 1e-1)

(1.0, 1.1102230246251565e-14, {'neval': 85766121})

2 mins 30 seconds  and 85 million evals is a long time and a lot of work to integrate a constant. :(

On the positive side, the answer is correct and the precision is a lot higher than asked for.

In [98]:
my_integrate(prob_all_sheep_see_each_other, 1e-1)

(0.027211883466817615, 0.014039624379012149, {'neval': 1812373437})

In [99]:
my_integrate(prob_all_sheep_see_each_other, 1e-2)

(0.027211883660186884, 0.009999827101293548, {'neval': 1814810865})

1e-1: (0.027211883466817615, 0.014039624379012149, {'neval': 1812373437}) , 110m

1e-2: (0.027211883660186884, 0.009999827101293548, {'neval': 1814810865}) , 105m


# Conclusion.

It seems pretty likely that the probability is 0.02721188 +/- epsilon.

That's tantalizingly close to but different from our reference of 0.029167